lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+//===- GrandCentral.cpp - Ingest black box sources --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Implement SiFive's Grand Central transform.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/StringSwitch.h"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Mutable store of information about an Element in an interface.  This is
+/// derived from information stored in the "elements" field of an
+/// "AugmentedBundleType".  This is updated as more information is known about
+/// an Element.
+struct ElementInfo {
+  /// Encodes the "tpe" of an element.  This is called "Kind" to avoid
+  /// overloading the meeaning of "Type" (which also conflicts with mlir::Type).
+  enum Kind {
+    Error = -1,
+    Ground,
+    Vector,
+    Bundle,
+    String,
+    Boolean,
+    Integer,
+    Double
+  };
```

Comment:
Maybe a name that contains "type" in some way could help with readability and indicate that this is actually a rough equivalent of our FIRRTL types? Something like `ElementType` maybe? Or actually, you could call it `Type` and, since it's nested in `ElementInfo`, this will end up being called `ElementInfo::Type`.

"Nit supreme", as @youngar puts it :wink:

Reply:
Nah, this is fair. @lattner was against `Type` as that's so overloaded and means something very specific. I opted for `Kind` where the name of the value is `tpe`. In the annotation it's just called `tpe` (which is to get around `type` being a keyword). I tried to clarify what is going on here in the comment.  Calling it `Tpe` is another option.

Reply:
I'm leaving this unchanged for now, but I'm open to further discussion.